### ISSUE TITLE
Delay before animation begins

### ISSUE THREAD

**Issue by Richardinho** [reactjs/react-transition-group#284]
In the first example given in the documentation there is a noticeable delay between clicking the button and the animation starting. This delay seems to correspond to the duration of the timeout property. Clearly, one would desire the animation to begin straight away, normally, as well as being able to configure a delay in some cases. It's not so obvious in the example because the timeout isn't very large, but when you set it to a bigger number you begin to see the issue.

In my own experiment I set start and end properties for the respective entering and exit states. 
```
const transitionStyles = {
  entering: { left: 0 },
  entered: { left: '100px' },
};
``` 
What then happens is that the start property is immediately set, but that then a delay occurs before the animation kicks in.

I get the desired effect if I set the property to be the same in both states (and set the default to 0), but this doesn't seem right.
```
const defaultStyles = {
  left: 0,
}
const transitionStyles = {
  entering: { left: '100px' },
  entered: { left: '100px' },
};
```
Here is the full code for the my experiment when there is an unwanted delay at the start of the animation:
```
import ReactDOM from 'react-dom';
import React from 'react';
import { Transition } from 'react-transition-group';

const duration = 1600;

const defaultStyle = {
  transition: `left ${duration}ms linear`,
  position: 'absolute',
  top: 200,
  width: '100px',
  height: '100px',
}

const transitionStyles = {
  entering: { left: '0' },
  entered:  { left: '400px'},
};

class Foo extends React.Component {
  constructor () {
    super();
    this.state = {
      in: false,
      duration
    };
    this.onClick = this.onClick.bind(this);
    this.onClickOff = this.onClickOff.bind(this);
  }

  onClick() {
    this.setState({
      in: true,
    });
  }

  onClickOff() {
    this.setState({
      in: false,
    });
  }

  render() {
   return (
     <div>
       <button onClick={this.onClick}>clickOn</button>
       <button onClick={this.onClickOff}>clickOff</button>
       <Transition
        onEntering={() => { console.log('on entering')}}
        onEntered={() => { console.log('on entered')}}
       in={this.state.in}
       timeout={this.state.duration}>
       {(state) => (
         <div style={{
           background: 'blue',
             ...defaultStyle,
             ...transitionStyles[state]
         }}>
         <div>
         <h2>This is Bar</h2>
         </div>
         </div>
       )}
       </Transition>
     </div>
   );
  }
}

ReactDOM.render(
  <Foo/>, document.getElementById('app')
);

```

**Comment by k-funk** [comment]
I had this same issue. I think that using `duration` for both `left ${duration}ms linear` and `timeout={duration}`. I solved the problem in my particular use case by setting `timeout={0}`.

**Comment by yaodingyd** [comment]
Change to:
```javascript
const transitionStyles = {
  entering: { left: '400px' },
  entered:  { left: '400px' },
};
```

And the delay would be gone.

**Comment by Godsenal** [comment]
I had the same issue. `timeout={0}` could fix it partially but sometimes enter transition doesn't work. 
I guess `timeout = {0}` makes `entered` state apply right after mounted before `entering` state apply to dom element.

**Comment by VincentNewkirk** [comment]
I'm running into the exact same issue. I've tried both setting ```timeout``` to 0 and setting the ```entering``` and ```exiting``` ```transitionStyles``` to the same value, but there's still a noticeable delay before the transition begins.

**Comment by jquense** [comment]
Timeout is the duration of the animation. Setting it to 0 isn't the right approach,even if it appears to work. Underneath the state is all out of wack. A delay in something starting is probably due an improperly set up css transition, or failure to reflow the browser at the appropriate time. For most folks we recommend using the CSSTransition instead of the lower level Transition component unless you're familiar with the various complexities around doing case transitions in js

**Comment by silvenon** [comment]
Hopefully this comment will shed some more light on this behavior: https://github.com/reactjs/react-transition-group/issues/307#issuecomment-377073024

**Comment by XcrossD** [comment]
It's good to know that `timeout` is actually the duration of the animation, but what do I do if I actually want a delay before the animation? For example, `delay` (don't do anything) for 500ms, then `timeout` (run animation) for 500ms.

**Comment by silvenon** [comment]
You can use `setTimeout` to delay whatever is setting the `in` prop to `true`.

**Comment by vidjuheffex** [comment]
@jquense it has? Just updated, same problem remains...

**Comment by silvenon** [comment]
It’s an automatic message. Let me know if there’s a specific example where it’s hard to make this delay with the current API.

**Comment by 1oginov** [comment]
Same problem, using v2.4.0

It's more convinient to to use `Transition` because it provides `state`. For example I made HOC:

```js
const withTransition = (inPredicate, timeout) => (BaseComponent) => {
  const factory = React.createFactory(BaseComponent);

  const WithTransition = props => (
    <Transition in={inPredicate(props)} timeout={timeout}>
      {transitionState => factory({ ...props, transitionState })}
    </Transition>
  );

  if (process.env.NODE_ENV !== 'production') {
    WithTransition.displayName = `withTransition(${BaseComponent.displayName})`;
  }

  return WithTransition;
};
```

And use it like so:

```js
const Root = ({ classes, transitionState }) => (
  <div className={classNames(classes.root, classes[transitionState])}>Root</div>
);
```

```js
withTransition(() => {/* some logic returns true to show component */}, 1000)(Root);
```

In this case we can use JSS and keep appearing logic, view and classes separated. Classes prop comes to `Root`, but not `withTransition` HOC.

P.S.: Anyway, I tried `CSSTransition`, it has the same delay before animation starts.

**Comment by silvenon** [comment]
Please open a new issue with a complete demo and description.

**Comment by 1oginov** [comment]
I'm not interested in, I described case when `Transition` preferred over `CSSTransition`. This is just a reminder that the issue still exists. The issue is already described above.

`timeout` prop adds a delay _before_ animation. We had to use `timeout={0}` as the only solution despite it's wrong approach :ok_woman:

**Comment by silvenon** [comment]
`CSSTransition` inherits **all** the properties of `Transition`, **including** state. See [our demo](http://reactcommunity.org/react-transition-group/css-transition) for an example of this. Basically:

```jsx
<CSSTransition in={inProp} timeout={1000} classNames="fade">
  {state => <div>{state}</div>}
</CSSTransition>
```

Animating the appearance is tricky. Unless you use `CSSTransition` or find another way to trigger a reflow yourself, there's going to be a delay. Let me know if using `CSSTransition` somehow isn't an option in your case.

**Comment by 1oginov** [comment]
Thanks @silvenon !

I tried `CSSTransition`, but the problem was because `.entering` should have the same style as `.entered` to start animation right after `state` changes. So I set the same style as guys above. I'll stick with `Transition` temporary until find some issues with this approach. 👍

**Comment by silvenon** [comment]
I'm a bit lost in the HOC logic above, but the `-active` classes are the ones you need to apply the transitions to. Anyway, I hope you managed to get it working without resorting to `timeout={0}`.

**Comment by 1oginov** [comment]
Yes, HOC above works with `timeout`. It passes `state` down to the component (as `transitionState` prop), so the component can use `state` value to apply different classes:

```js
const Root = ({ classes, transitionState }) => (
  <div className={classNames(classes.root, classes[transitionState])}>Root</div>
);
```

As I can understand, `state` can be `entering`, `entered`, `exiting` and `exited`... The idea is to specify classes in the component and to have transition and other logic seprated.

**Comment by silvenon** [comment]
You shouldn't apply classes using the state, `CSSTransition` manages classes on its own, you should use those. I know that's not in the spirit of CSS-in-JS, but ATM you won't be able to achieve the desired result any other way.

I'm open to suggestions for a more flexible API.

**Comment by 1oginov** [comment]
That's why I'm not using `CSSTransition` :)

`Transition` is enough, it works fine, what the problem?

**Comment by silvenon** [comment]
You said there was a problem:

> Same problem, using v2.4.0

I was just trying to understand what it is.

**Comment by timhecker** [comment]
Any news about the delay on animations? I'm using a `<CSSTransition>` inside another one and I would that the second one starts with a delay.

@silvenon temporary can you make an example with your solution: 

> You can use setTimeout to delay whatever is setting the in prop to true.

Thanks

**Comment by silvenon** [comment]
I haven't tested this, but here's a (possibly naive) example:

```jsx
import React from 'react'
import { CSSTransition } from 'react-transition-group'

const DELAY = 1000
const DURATION = 500

class MyComponent extends React.Component {
  state = {
    inProp: false,
  }
  
  timerId = null

  render() {
    const { inProp } = this.state;
    return (
      <div>
        <CSSTransition
          in={inProp}
          timeout={DURATION}
          classNames="fade"
        >
          {/* ... */}
        </CSSTransition>
        <button
          type="button"
          onClick={() => {
            this.timerId = setTimeout(() => {
              this.setState({
                inProp: true,
              })
            }, DELAY)
          }}
        >

        </button>
      </div>
    )
  }
}
```

Make sure to stop the timer with `clearTimeout(this.timerId)` when necessary, depending on your use case.

**Comment by Mati365** [comment]
Any changes? 2.4.0 still have timeout before starting anim(as @1oginov said).

**Comment by silvenon** [comment]
This issue talks about two opposite things, to remove the delay and to add it. I provided the solutions to both, so please share some code and I'll help you get rid of the initial delay.

**Comment by std4453** [comment]
Any updates on this thread?
I need children `TransitionGroup` to first exit and then enter, not wanting them to happen simultaneously.
I can set delay in `CSSTransition` manually but a) it's ugly b) it will introduce a delay if `appear` is set.
So is there a better solution? Or should I open an new issue?

**Comment by rppld** [comment]
The example in [this comment](https://github.com/reactjs/react-transition-group/issues/223#issuecomment-334748429) solved it for me.

**Comment by silvenon** [comment]
@std4453 so you're trying to **add** a delay? Perhaps best to open a new issue with a demo and an detailed description of what you're trying to achieve.

**Comment by talham7391** [comment]
To those still having this issue: I was having the same issue and this is how I fixed it.

The documentation at: https://reactcommunity.org/react-transition-group/transition suggests that you should put your starting css styles in the "entering" state and your target css styles in the "entered" state. This is wrong, doing it this way causes a delay and heres why. Suppose we have the following setup:

defaultState: opacity: 0; transition: opacity 500ms;
enteringState: opacity: 0;
enteredState: opacity 1;
timeout = 500

When this animation is played, the first 500 milliseconds "enteringState" will be applied. Note that there is no difference between the entering state and the default, thus nothing happens and this is seen as the "delay". After 500ms, the "enteredState" is applied and you will then see an animation because the "transition" css property was set to 500ms.

The reason why setting timeout=0 solved the problem for a lot of people could have been because with timeout set to 0, the "enteredState" gets applied right away and the CSS transition happens. This is wrong as the CSS animation and the Transition group timeout aren't lining up.

Solution:

The setup above should have been:

defaultState: opacity: 0; transition: opacity 500ms;
enteringState: opacity: 1;    <-- this is the only thing that changed.
enteredState: opacity 1;
timeout = 500

Now when "enteringState" is applied, the CSS transition will begin to happen right away and it will lineup with the Transition timeout. As a note, your "enteringState" & "enteredState" (and your "exitingState" & "exitedState") should be the same.

**Comment by silvenon** [comment]
@talham7391 thanks for pointing this out, I wasn't aware of this problem in the docs, now this has been fixed. ✅

**Comment by jwm0** [comment]
@talham7391 entering state does not trigger the transition for me. The workaround I'm using right now is to set duration to { enter: 0, exit: duration }

**Comment by Mati365** [comment]
@jwm0 if you pass 0 to enter value it is hightly possible that enter phrase will be not animated. I have tried that and sometimes component had appeared without any animation - try to set 60ms instead of 0. Browser engine will be able to set transitions correctly (without batching etc).

**Comment by jwm0** [comment]
@Mati365 it feels hacky but I'll look into it - let me know if you remember any particular scenarios where it would not animate. Also, can you confirm that the solution @talham7391 provided also doesn't seem to be working for you?

**Comment by rmolinamir** [comment]
This is how I fixed this, use `transition-delay` property in your `enter-active` and `exit-active`, that is not all you have to do, however.

The prop `timeout`, is the **total** duration of the animation, delay included, so if you're going to add the css `transition` property's duration, make sure the sum of the *delay* and the *duration* of the transition is equal to the `timeout` prop, for example:

```
.${classNames}-enter {
      opacity: 0;
      transform: scale(0.95);
    }
    .${classNames}-enter-active {
      opacity: 1;
      transform: scale(1);
      transition: opacity ${duration}ms, transform ${duration}ms;
      transition-delay: ${delay}ms;
    }
```

Where `timeout` would be equal to the sum of `duration` and `delay`. In my component, it looks like this:

```
  const delay = animationTimeout * 0.33 * animationMultiplier * animationDelayMultiplier;
  const duration = animationTimeout * 0.66 * animationMultiplier;

  ...

  return (
      <CSSTransition
        in={in}
        timeout={duration + delay}
        classNames={classNames}
        onEnter={onEnter}
        onExited={onExited}
        unmountOnExit={unmountOnExit}
      >
        ...
```

And here's an example of how it looks in production: https://i.gyazo.com/23172bc8d4a81ce03389e4c7da7dd7ee.mp4

**Comment by littlemilkstudio** [comment]
I had an issue getting the `defaultState`  to render with @talham7391 solution, but agree that it looks correct.

With @jwm0 proposed solution of setting `timeout={{ enter: 0, exit: duration }}` I ran into the issue where sometimes the enter phase wasn't animated.

To get both working I created an interpretation layer between the `Transition` and `children`. Seemed to fix both issues while staying true to the model @talham7391 proposed.

```jsx
const entering = 'entering';
const entered = 'entered';
const exiting = 'exiting';
const exited = 'exited';

const machine = {
  initial: 'default',
  states: {
    default: { entering },
    entering: { entered },
    entered: { exiting },
    exiting: { exited },
    exited: {},
  },
};

const reducer = (state, action) => machine.states[state]?.[action] || state;

// Used as an interpretation layer between Transition
// and children being called with state. Allows for first
// render of default styles while keeping timeout in sync
// with state.
export function Interpreter({ state, children }) {
  const [scheduledState, dispatch] = useReducer(
    reducer,
    state === entered ? state : machine.initial
  );
  const [nextRender, setNextRender] = useState(state);

  // Ensures we get a paint with every current state
  // we are on before we interpret an update.
  useEffect(() => {
    const id = requestAnimationFrame(() => setNextRender(state));
    return () => cancelAnimationFrame(id);
  }, [state]);

  // Use our state machine implementation to deterministically order state
  // transitions. Ensures following order & at least one render on each state:
  // default -> entering -> entered -> exiting -> exited
  useEffect(() => {
    dispatch(nextRender);
  }, [nextRender]);

  // We want to opt out of our expected behavior immediately if there's a premature
  // enter or exit scenario.
  //
  // It's important that these state changes are reflected on exactly this render,
  // as opposed to in the state machine, to prevent content flash where two children
  // of a transition group can both be in an `enter` or `exit` state in the same paint
  // when only one is entering and the other is exiting.
  const isPrematureExit =
    [entering, entered].includes(scheduledState) && state === 'exiting';
  const isPrematureEntrance =
    [exiting].includes(scheduledState) && [entering, entered].includes(state);
  const shouldUseBaseState = isPrematureEntrance || isPrematureExit;
  return children(shouldUseBaseState ? state : scheduledState);
}

export function ScheduledTransition({ children, ...props }) {
  return (
    <Transition {...props}>
      {(state) => (
        <Interpreter state={state}>
          {(scheduledState) => children(scheduledState)}
        </Interpreter>
      )}
    </Transition>
  );
}
```


Allowed me to use `ScheduledTransition` as a drop in replacement for `Transition` while solving all the problems discussed in this thread:
```jsx
const duration = 300;

const transitionStyles = {
  entering: `transform: none;`,
  entered:  `transform: none;`,
  exiting:  ' transform: translateX(100%);',
  exited:  ' transform: translateX(100%);',
};

const Controller = styled.div`
  // default styles
  transform: translateX(-100%);
  transition: transform ${duration}ms ease-out;
  
  // Transition Styles
  ${({ state }) => transitionStyles[state]}
`;

const Fade = ({ in: inProp }) => (
  <ScheduledTransition in={inProp} timeout={duration} mountOnEnter unmountOnExit>
    {state => (
      <Controller state={state}>
        I'm a translate Transition!
      </Controller>
    )}
  </ScheduledTransition>
);
```

Please let me know if I'm misusing the package at all and there's no need for this middle layer. However, thought i'd mention here in case someone is running into the same issue because it seemed to fix it for me and provide reliable behavior
